src/lib.rs

```diff
@@ -51,144 +49,406 @@
+    /// which have been granted for this `access_token`
+    /// You could use macro [scopes!](crate::scopes) to build it at compile time easily
+    #[serde(default, with = "space_separated_scope")]
+    pub scope: HashSet<String>,
+}
+
+impl Default for Token {
+    fn default() -> Self {
+        Token {
+            access_token: String::new(),
+            expires_in: Duration::seconds(0),
+            expires_at: Some(Utc::now()),
+            refresh_token: None,
+            scope: HashSet::new(),
+        }
+    }
+}
+
+impl Token {
+    /// Tries to initialize the token from a cache file.
+    // TODO: maybe ClientResult for these things instead?
+    pub fn from_cache<T: AsRef<Path>>(path: T) -> Option<Self> {
+        let mut file = fs::File::open(path).ok()?;
+        let mut tok_str = String::new();
+        file.read_to_string(&mut tok_str).ok()?;
+
+        serde_json::from_str::<Token>(&tok_str).ok()
+    }
+
+    /// Saves the token information into its cache file.
+    pub fn write_cache<T: AsRef<Path>>(&self, path: T) -> ClientResult<()> {
+        let token_info = serde_json::to_string(&self)?;
+
+        let mut file = fs::OpenOptions::new().write(true).create(true).open(path)?;
+        file.set_len(0)?;
+        file.write_all(token_info.as_bytes())?;
+
+        Ok(())
+    }
+
+    /// Check if the token is expired
+    pub fn is_expired(&self) -> bool {
+        self.expires_at
+            .map_or(true, |x| Utc::now().timestamp() > x.timestamp())
+    }
+}
+
+/// Simple client credentials object for Spotify.
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct Credentials {
+    pub id: String,
+    pub secret: String,
+}
+
+impl Credentials {
+    pub fn new(id: &str, secret: &str) -> Self {
+        Credentials {
+            id: id.to_owned(),
+            secret: secret.to_owned(),
+        }
+    }
+
+    /// Parses the credentials from the environment variables
+    /// `RSPOTIFY_CLIENT_ID` and `RSPOTIFY_CLIENT_SECRET`. You can optionally
+    /// activate the `env-file` feature in order to read these variables from
+    /// a `.env` file.
+    pub fn from_env() -> Option<Self> {
+        #[cfg(feature = "env-file")]
+        {
+            dotenv::dotenv().ok();
+        }
+
+        Some(Credentials {
+            id: env::var("RSPOTIFY_CLIENT_ID").ok()?,
+            secret: env::var("RSPOTIFY_CLIENT_SECRET").ok()?,
+        })
+    }
+}
+
+/// Structure that holds the required information for requests with OAuth.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct OAuth {
+    pub redirect_uri: String,
+    /// The state is generated by default, as suggested by the OAuth2 spec:
+    /// [Cross-Site Request Forgery](https://tools.ietf.org/html/rfc6749#section-10.12)
+    pub state: String,
+    /// You could use macro [scopes!](crate::scopes) to build it at compile time easily
+    pub scope: HashSet<String>,
```

Comment:
Perhaps, we should rename it as `scopes`?

Reply:
Yea I prefer that as well.